### ISSUE TITLE
Setup Invite System

### ISSUE THREAD

**Issue by dave-mills** [stats4sd/tape-data-system#3]
We need to be able to send invites via email for users to join the platform. We need different types of invite:

- Invitation to join a specific team. We should be able to send an invitation to any team through the admin panel and to the current team through the app panel. 
- Invitation to join the platform with a specific user role (currently, just "admin".) This should be done through the admin panel. 

We have this type of invite system in other platforms, e.g.:
- [MAP Project](https://github.com/stats4sd/tape-giz)

I believe in that platform the functionality is split a bit between the main app and the [Filament ODK Link](In our other platforms, the invite system is split in functionality between the main app and the [filament-odk-link](https://github.com/stats4sd/filament-odk-link) package, meaning it is a bit of a pain to setup. ) package, so it might be a bit of a pain to set up just by copying it.

**Comment by dave-mills** [comment]
Current status: The system sends out an email, but the link in the email doesn't work. 

I also don't think the registration page handles the invite tokens correctly (in fact, I'm pretty sure it doesn't because it will be using the default Filament registration).

**Comment by dan-tang-ssd** [comment]
> We need to be able to send invites via email for users to join the platform. We need different types of invite:
> 
>     * Invitation to join a specific team. We should be able to send an invitation to any team through the admin panel and to the current team through the app panel.
> 
>     * Invitation to join the platform with a specific user role (currently, just "admin".) This should be done through the admin panel.
> 
> 
> We have this type of invite system in other platforms, e.g.:
> 
>     * [MAP Project](https://github.com/stats4sd/tape-giz)
> 
> 
> I believe in that platform the functionality is split a bit between the main app and the [Filament ODK Link](In our other platforms, the invite system is split in functionality between the main app and the [filament-odk-link](https://github.com/stats4sd/filament-odk-link) package, meaning it is a bit of a pain to setup. ) package, so it might be a bit of a pain to set up just by copying it.

I did a preliminary study in tape-giz. 
Here are my findings:

User page location: 
 - Login > System Admin > Users CRUD

User page list view:
 - name, email, teams, roles, Edit button
 - search feature
 - select records > Bulk actions: delete selected

Invite user button, modal popup
 - Email address
 - select role(s)
 - select team
   - can select a team only if role "Country Team" is ticked

Related program source code:
 - App\Filament\Resources\UserResources
 - App\Models\User, implements FilamentUser

Program trigger to send email
 - App\Filament\Resources\UserResource\Pages\ListUsers

---

Screen shots:

![image](https://github.com/user-attachments/assets/89d98534-a543-4a93-b267-81a11665b098)

![image](https://github.com/user-attachments/assets/f1a8e4fc-3c22-442b-82d0-ebef6b827828)

![image](https://github.com/user-attachments/assets/ea710dde-369b-4a4b-b7c4-ece8249fa389)

![image](https://github.com/user-attachments/assets/eab3a4d2-307b-4fe2-ad75-ff6c3aabbfa9)

**Comment by dan-tang-ssd** [comment]
> Invitation to join a specific team. We should be able to send an invitation to any team through the admin panel and to the current team through the app panel.

In admin panel, user can select any team for sending invitation. 
In app panel, user cannot select team, the team is pre-selected to the user's team and the team selection box should be disabled.

> Invitation to join the platform with a specific user role (currently, just "admin".) This should be done through the admin panel.

In this cause, every user will be admin, all of them can access admin panel User resource. 
Any user will be able to send invitation to any team.
Um... maybe we need more roles so that admin panel User resource can be accessed by system admin only...

> I believe in that platform the functionality is split a bit between the main app and the [Filament ODK Link](In our other platforms, the invite system is split in functionality between the main app and the [filament-odk-link](https://github.com/stats4sd/filament-odk-link) package, meaning it is a bit of a pain to setup. ) package, so it might be a bit of a pain to set up just by copying it.

Um... Are you referring to business logic for retrieving ODK submission in filament-odk-link package? I am not sure whether there are filament related code (for front end) is in the package...

**Comment by dan-tang-ssd** [comment]
> Current status: The system sends out an email, but the link in the email doesn't work.
> 
> I also don't think the registration page handles the invite tokens correctly (in fact, I'm pretty sure it doesn't because it will be using the default Filament registration).

Um... do you mean the filament registration page which is not enabled now? As I would like to find it out and give it a try to see how it does...

**Comment by dan-tang-ssd** [comment]
> > Current status: The system sends out an email, but the link in the email doesn't work.
> > I also don't think the registration page handles the invite tokens correctly (in fact, I'm pretty sure it doesn't because it will be using the default Filament registration).
> 
> Um... do you mean the filament registration page which is not enabled now? As I would like to find it out and give it a try to see how it does...

Oh... I found where to sign up. It is in the login page of tape-data-system rather than tape-giz.

I can sign up successfully. But I have not received any email yet, maybe mailgun is not setup properly in my local env yet.

---

Screen shots:

![image](https://github.com/user-attachments/assets/bcef86b7-12d7-49f0-b0b6-e5738a598030)

![image](https://github.com/user-attachments/assets/fd371fec-6da0-409b-8a7d-d29c245dd0ab)

![image](https://github.com/user-attachments/assets/14ba6569-ec90-4a1c-b557-64e572f5f310)

**Comment by dave-mills** [comment]
> In admin panel, user can select any team for sending invitation.
> In app panel, user cannot select team, the team is pre-selected to the user's team and the team selection box should be disabled.

Exactly - or better still we don't even show a team selection box - the App Panel has multi-tenancy, so you'll always be in the context of a specific team. We don't need any team select field in a form to invite new users. 

> In this cause, every user will be admin, all of them can access admin panel User resource.
> Um... maybe we need more roles so that admin panel User resource can be accessed by system admin only...

Yes. The key point is that the "role invites" means the new user will have a site-wide role. This can be *any* site-wide role. For the TAPE system we currently only have admin, but later I can imagine we'll have things like "Global Reviewer", "Team Manager", etc etc. The role invite system should work the same no matter what roles the system has. 

> Um... Are you referring to business logic for retrieving ODK submission in filament-odk-link package? I am not sure whether there are filament related code (for front end) is in the package...

No. Take a look at the package models inside the `Stats4SD\FilamentOdkLink\Models\TeamManagement` namespace. The RoleInvite and TeamInvite models are the actual invites, and the `Team` model there has the methods and relationships to handle.

**Comment by dave-mills** [comment]
> I can sign up successfully. But I have not received any email yet, maybe mailgun is not setup properly in my local env yet.

Exactly. The MAP system uses the older Backpack panel to login (because we added the Filament panel for data and results half-way through). 

Re email not being sent, yep - as I said: https://github.com/stats4sd/tape-data-system/issues/3#issuecomment-2088438275

**Comment by dave-mills** [comment]
I wonder if it would be better long term to extend the Filament panels in the package (like the `TeamResource`) to present a default front-end, at least for the "Admin" panel of our systems. Similar to the Xlsform handling, where we start with a default working set-up, and can then customise it for specific systems if needed.


One other thing:  By default, we should assume our systems are invite-only, so registration should only work if a user has a valid invite token. (Which means either a TeamInvite or RoleInvite token where `is_confirmed === false`, and they should be required to use that invite's email address).

**Comment by dan-tang-ssd** [comment]
> I wonder if it would be better long term to extend the Filament panels in the package (like the `TeamResource`) to present a default front-end, at least for the "Admin" panel of our systems. Similar to the Xlsform handling, where we start with a default working set-up, and can then customise it for specific systems if needed.

Yes. It is good to have a Users page in the package to manage users, and send invites to any team.
Maybe we can add it below Teams page.

![image](https://github.com/user-attachments/assets/9ce5c1bd-a292-457a-9989-35b8281d6f12)


> One other thing: By default, we should assume our systems are invite-only, so registration should only work if a user has a valid invite token. (Which means either a TeamInvite or RoleInvite token where `is_confirmed === false`, and they should be required to use that invite's email address).

Yes, we should disable the "sign up" feature in login page.

![image](https://github.com/user-attachments/assets/6af8003e-8b8c-47f5-9379-63354bb1e059)
